File: server/src/game/Physics.ts
export interface AABB {
  x: number;
  y: number;
  w: number;
  h: number;
}

export interface Circle {
  x: number;
  y: number;
  r: number;
}

export function aabbOverlap(a: AABB, b: AABB): boolean {
  return a.x < b.x + b.w && a.x + a.w > b.x && a.y < b.y + b.h && a.y + a.h > b.y;
}

export function entityAABB(x: number, y: number, w: number, h: number): AABB {
  return { x: x - w / 2, y: y - h / 2, w, h };
}

export function entityCircle(x: number, y: number, r: number): Circle {
  return { x, y, r };
}

export function circleOverlap(a: Circle, b: Circle): boolean {
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  const rr = a.r + b.r;
  return dx * dx + dy * dy <= rr * rr;
}

export function circleAabbOverlap(circle: Circle, box: AABB): boolean {
  const nearestX = clamp(circle.x, box.x, box.x + box.w);
  const nearestY = clamp(circle.y, box.y, box.y + box.h);
  const dx = circle.x - nearestX;
  const dy = circle.y - nearestY;
  return dx * dx + dy * dy <= circle.r * circle.r;
}

export function distanceSquared(x1: number, y1: number, x2: number, y2: number): number {
  const dx = x2 - x1;
  const dy = y2 - y1;
  return dx * dx + dy * dy;
}

export function distance(x1: number, y1: number, x2: number, y2: number): number {
  return Math.sqrt(distanceSquared(x1, y1, x2, y2));
}

export function clamp(value: number, min: number, max: number): number {
  return Math.max(min, Math.min(max, value));
}

export function isInSafeZone(
  x: number,
  y: number,
  spawnX: number,
  spawnY: number,
  safeRadius: number
): boolean {
  return distanceSquared(x, y, spawnX, spawnY) <= safeRadius * safeRadius;
}

File: server_go/internal/domain/physics/physics.go
// Package physics provides geometric primitives shared by the simulation
// systems: axis-aligned bounding boxes, circles, distance helpers, and the
// snapshot quantization rule.
package physics

import (
	"math"

	"github.com/williamisnotdefined/zelda-proto/server_go/internal/domain/world"
)

// AABB is an axis-aligned bounding box anchored at its top-left corner.
type AABB struct {
	X, Y, W, H float64
}

// Circle is a 2D circle defined by its center and radius.
type Circle struct {
	X, Y, R float64
}

// EntityAABB returns the bounding box centered on (x, y) with size w x h.
func EntityAABB(x, y, w, h float64) AABB {
	return AABB{X: x - w/2, Y: y - h/2, W: w, H: h}
}

// EntityCircle returns a circle centered on (x, y) with radius r.
func EntityCircle(x, y, r float64) Circle {
	return Circle{X: x, Y: y, R: r}
}

// AABBOverlap reports whether two AABBs overlap (open intervals on both axes).
func AABBOverlap(a, b AABB) bool {
	return a.X < b.X+b.W && a.X+a.W > b.X && a.Y < b.Y+b.H && a.Y+a.H > b.Y
}

// CircleOverlap reports whether two circles overlap.
func CircleOverlap(a, b Circle) bool {
	dx := b.X - a.X
	dy := b.Y - a.Y
	rr := a.R + b.R
	return dx*dx+dy*dy <= rr*rr
}

// CircleAABBOverlap reports whether circle c overlaps box b.
func CircleAABBOverlap(c Circle, b AABB) bool {
	nearestX := Clamp(c.X, b.X, b.X+b.W)
	nearestY := Clamp(c.Y, b.Y, b.Y+b.H)
	dx := c.X - nearestX
	dy := c.Y - nearestY
	return dx*dx+dy*dy <= c.R*c.R
}

// DistanceSquared returns the squared Euclidean distance between two points.
func DistanceSquared(x1, y1, x2, y2 float64) float64 {
	dx := x2 - x1
	dy := y2 - y1
	return dx*dx + dy*dy
}

// Distance returns the Euclidean distance between two points.
func Distance(x1, y1, x2, y2 float64) float64 {
	return math.Sqrt(DistanceSquared(x1, y1, x2, y2))
}

// Clamp restricts value to the inclusive range [min, max]. When min > max the
// behaviour is undefined; callers must validate inputs.
func Clamp(value, minimum, maximum float64) float64 {
	if value < minimum {
		return minimum
	}
	if value > maximum {
		return maximum
	}
	return value
}

// IsInSafeZone reports whether (x, y) lies within the spawn safe zone defined
// by (spawnX, spawnY, safeRadius).
func IsInSafeZone(x, y, spawnX, spawnY, safeRadius float64) bool {
	return DistanceSquared(x, y, spawnX, spawnY) <= safeRadius*safeRadius
}

// QuantizePosition rounds a world-space coordinate to the snapshot precision
// used by both server and client (matches the legacy TS implementation).
func QuantizePosition(value float64) float64 {
	return math.Round(value*world.PositionPrecision) / world.PositionPrecision
}
